### PR TITLE
Identity Refactor

### PR DESCRIPTION
**Pull request by saneshark**
## Description of change
Identified several issues that this PR resolves;
1. URLService logic was incorrect. It was using this concept of real_authn_context which has the unintended consequence of setting the wrong link.
2. Removed all notion of "real_authn_context" and sticking with authn_context, but needed to continue to surface authn_context to the FE in the same manner as they expect, so added custom code in the user serializer and specs to test out that out granularly. They should transition to using sign_in instead.
3. With real_authn_context removed, some additional refactoring was needed elsewhere, especially some handling of nil LOA cases. It appears that for multifactor endpoints we were not able to determine the LOA reliably from the already existing type. Making this improvement demanded rewriting a lot of specs.
4. introduced ResponseBuilder as a helper that is shared between a bunch of different specs for mocking out RubySaml responses.
5. Other minor changes and improvements for readability and more consistent Raven.extra_context and tag management.

## Testing done
Specs and Local testing only.

## Testing planned
Thorough regression testing of as many identity flows as we can do.

#### Unique to this PR

- [ ] _Replace this line with individual tasks unique to your PR_

#### Applies to all PRs

- [ ] Appropriate logging
- [ ] Swagger docs have been updated, if applicable
- [ ] Provide link to originating GitHub issue, or connected to it via ZenHub
- [ ] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [ ] Provide which alerts would indicate a problem with this functionality (if applicable)

### CODE REVIEW

File: lib/saml/user.rb [hunk 1]
```diff
@@ -10,113 +10,89 @@ module SAML
...
     # returns the attributes that are defined below, could be from one of 3 distinct policies, each having different
     # saml responses, hence this weird decorating mechanism, needs improved abstraction to be less weird.
     def serializable_attributes
       %i[authn_context]
     end
 
-    def dslogon?
-      saml_attributes.to_h.keys.include?('dslogon_uuid')
-    end
+    # NOTE: The actual exception, if any, should get raised when to_hash is called. Hence "suppress"
```
Comment (annaswims) [inline lib/saml/user.rb]: it looks like you took out `suppress(Exception)` but mention it in the comment.

File: lib/saml/user.rb [hunk 2]
```diff
@@ -10,113 +10,89 @@ module SAML
...
     # should eventually have a special case for multifactor policy and refactor all of this
     # but session controller refactor is premature and can't handle it right now.
     def user_attributes_class
       case authn_context
       when 'myhealthevet'; then SAML::UserAttributes::MHV
       when 'dslogon'; then SAML::UserAttributes::DSLogon
-      else
+      when 'myhealthevet_multifactor', 'dslogon_multifactor', 'multifactor'
         SAML::UserAttributes::IdMe
+      when 'dslogon_loa3', 'myhealthevet_loa3', LOA::IDME_LOA3
```
Comment (annaswims) [inline lib/saml/user.rb]: simplify to?
````ruby 
      when 'myhealthevet_multifactor', 'dslogon_multifactor', 'multifactor',
            'dslogon_loa3', 'myhealthevet_loa3', LOA::IDME_LOA3,
             LOA::IDME_LOA1
        SAML::UserAttributes::IdMe
````

File: lib/saml/user.rb [hunk 3]
```diff
@@ -10,113 +10,89 @@ module SAML
...
-        end
+        log_message_to_sentry('SAML RESPONSE WARNINGS', :warn, warning_context)
       end
     end
 
-    # will be one of [loa1, loa3, multifactor, dslogon, mhv]
+    # will be one of AUTHN_CONTEXTS.keys
     # this is the real authn-context returned in the response without the use of heuristics
-    def real_authn_context
+    def authn_context
```
[the remark this answers is not in the document]
Comment (saneshark) [reply lib/saml/user.rb]: good point... hmmm... let me see if I can make that change easily.

File: lib/saml/user.rb (hunk 2 above)
Comment (saneshark) [reply lib/saml/user.rb]: yes, the line was too long and I didn't like the way it looked with the when clause spanning multiple lines. 😂

File: lib/saml/user.rb (hunk 1 above)
Comment (saneshark) [reply lib/saml/user.rb]: fixed this

File: lib/saml/user.rb [hunk 4]
```diff
@@ -10,113 +10,89 @@ module SAML
...
-          }
-          log_message_to_sentry("Issues in SAML Response - #{real_authn_context}", :warn, warning_context)
-        end
+        log_message_to_sentry('SAML RESPONSE WARNINGS', :warn, warning_context)
       end
     end
 
-    # will be one of [loa1, loa3, multifactor, dslogon, mhv]
+    # will be one of AUTHN_CONTEXTS.keys
     # this is the real authn-context returned in the response without the use of heuristics
```
[the remark this answers is not in the document]
Comment (saneshark) [reply lib/saml/user.rb]: fixed this

File: lib/saml/user.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (saneshark) [reply lib/saml/user.rb]: lets make this improvement in another PR. I tried to fix it, but it breaks a lot of specs and don't have time to fix it.

File: lib/saml/user.rb [hunk 5]
```diff
@@ -10,113 +10,89 @@ module SAML
...
     attr_reader :saml_response, :saml_attributes, :user_attributes
 
     def initialize(saml_response)
       @saml_response = saml_response
       @saml_attributes = saml_response.attributes
-      @user_attributes = user_attributes_class.new(saml_attributes, real_authn_context)
-      log_warnings_to_sentry!
+
+      Raven.extra_context(
+        saml_attributes: saml_attributes&.to_h,
```
[the remark this answers is not in the document]
Comment (saneshark) [reply lib/saml/user.rb]: added these as well as dslogon_value, which is ssn

**Comment by annaswims** [comment]
LGTM

**Pull request merged**